csharp/ql/src/semmle/code/csharp/Conversion.qll

```diff
@@ -78,125 +92,213 @@ private Type getTypeArgument(UnboundGenericType ugt, ConstructedType ct, int i,
+  }
 
-pragma [noinline]
-private predicate convIdentityStrictArrayTypeJoin(IdentityConvertibleArrayType fromType, IdentityConvertibleArrayType toType, int dim, int rnk) {
-  convIdentityStrictArrayElementType(fromType, toType.getElementType(), dim, rnk)
-}
+  private predicate convIdentityStrictArrayElementType(IdentityConvertibleArrayType fromType, ArrayElementType aet, int dim, int rnk) {
+    convIdentityStrict(fromType.getElementType(), aet) and
+    dim = fromType.getDimension() and
+    rnk = fromType.getRank()
+  }
 
-private predicate convIdentityStrictArrayElementType(IdentityConvertibleArrayType fromType, ArrayElementType aet, int dim, int rnk) {
-  convIdentityStrict(fromType.getElementType(), aet) and
-  dim = fromType.getDimension() and
-  rnk = fromType.getRank()
-}
+  /**
+   * Gets the number of different type arguments supplied for the type
+   * parameter at index `i` in unbound generic type `ugt`.
+   */
+  private int getTypeArgumentCount(UnboundGenericType ugt, int i) {
+    result = strictcount(Type arg |
+      exists(IdentityConvertibleConstructedType ct |
+        ct.getUnboundGeneric() = ugt |
+        arg = ct.getTypeArgument(i)
+      )
+    )
+  }
 
-/** A type that is an element type of an array type. */
-private class ArrayElementType extends Type {
-  ArrayElementType() {
-    this = any(ArrayType at).getElementType()
+  private int rnk(UnboundGenericType ugt, int i) {
+    result = rank[i + 1](int j, int k |
+      j = getTypeArgumentCount(ugt, k) |
+      k order by j, k
+    )
   }
-}
 
-private predicate convIdentityStrictConstructedType(IdentityConvertibleConstructedType fromType, IdentityConvertibleConstructedType toType) {
-  /* Semantically equivalent with
-   * ```
-   * ugt = fromType.getUnboundGeneric()
-   * and
-   * forex(int i |
-   *   i in [0 .. ugt.getNumberOfTypeParameters() - 1] |
-   *   exists(Type t1, Type t2 |
-   *     t1 = getTypeArgument(ugt, fromType, i, _) and
-   *     t2 = getTypeArgument(ugt, toType, i, _) |
-   *     convIdentity(t1, t2)
-   *   )
-   * )
-   * ```
-   * but performance is improved by explicitly evaluating the `i`th argument
-   * only when all preceding arguments are convertible.
+  private Type getTypeArgumentRanked(UnboundGenericType ugt, IdentityConvertibleConstructedType t, int i) {
+    result = getTypeArgument(ugt, t, rnk(ugt, i), _)
+  }
+
+  /**
+   * Holds if `fromTypeArgument` is identity convertible to `toTypeArgument`, and
+   * both types are the `i`th type argument in _some_ constructed type.
    */
-  exists(UnboundGenericType ugt |
-    convIdentityStrictConstructedTypeFromZero(ugt, fromType, toType, ugt.getNumberOfTypeParameters() - 1)
-  )
-}
+  pragma[nomagic]
+  private predicate convTypeArguments(Type fromTypeArgument, Type toTypeArgument, int i) {
+    exists(int j |
+      fromTypeArgument = getTypeArgumentRanked(_, _, i) and
+      toTypeArgument = getTypeArgumentRanked(_, _, j) and
+      i <= j and j <= i
```

Comment:
Why was `j` introduced? (I guess it's an optimizer workaround but is it still needed, and what's it doing?) Perhaps a doc and raise a ticket?

Reply:
Yes, to avoid the optimizer joining on `i`.